### ISSUE TITLE
More complex mathematical computations

### ISSUE THREAD

**Issue by cooljeanius** [cooljeanius/Son_Of_The_Black_Eye_Easy_Edition#2]
- [ ] #24
- [ ] #25
- [x] Silent Forest: come up with a new formula to determine how much gold the elf you interrogate carries
- [ ] #26

**Comment by cooljeanius** [comment]
Hey @celticminstrel you're generally pretty good at coming up with formulae; any help here? (We might have actually already discussed some of these on Discord, but I forget what conclusion we came to)

**Comment by CelticMinstrel** [comment]
Based solely on what you've said so far in the thread, it seems pretty vague what you're going for, so I'm not sure.

**Comment by cooljeanius** [comment]
> Based solely on what you've said so far in the thread, it seems pretty vague what you're going for, so I'm not sure.

Well ok let me try rephrasing in the form of questions:
1. (actually I'll skip this one for now)
2. It's possible to make the value of `{LIMIT_CONTEMPORANEOUS_RECRUITS}` dynamic, i.e., so it depends on a variable that can update throughout the game, correct? When I asked previously, it seemed that that was the case, but I'm still not entirely sure if it'll work... Also how would I write a formula to count the number of units with fire-type attacks that a certain side has? Originally asked here: https://discord.com/channels/231976805987385345/442775044590927873/752589288092467222; I asked a related question here: https://discord.com/channels/231976805987385345/442775044590927873/793326799786344488 but forget what I did with that info... Screenshots: 
<img width="1015" alt="Screen Shot 2021-05-30 at 16 37 52" src="https://user-images.githubusercontent.com/1388660/120119480-df786880-c165-11eb-8b41-dd1d380a8535.png">
<img width="1032" alt="Screen Shot 2021-05-30 at 16 37 01" src="https://user-images.githubusercontent.com/1388660/120119483-e0a99580-c165-11eb-9a55-6af69820dff4.png">
3. I have an event where, when a unit dies, it drops some gold, and I want to change how that amount of gold is calculated. The code (with comment) currently looks like this:

```
        [gold]
            side=1
            # TODO: maybe get a bit more fancy with the amounts here.
            # ideas: MAX(unit.cost, amount), MIN(unit.cost, amount),
            # the average of [all units recruitable by|active on] [the unit's side|both elvish sides],
            # something else? (not sure how to actually calculate those)
            amount={ON_DIFFICULTY 60 40 20}
        [/gold]
```

Which of those ideas do you think would work best, and how would I implement it?
4. I found where we discussed this one previously, it was here: https://www.wesnoth.org/irclogs/2020/10/%23wesnoth-umc-dev.2020-10-22.log ...so yeah, how should I proceed from there?

**Comment by CelticMinstrel** [comment]
2. The recruit amount parameter to `{LIMIT_CONTEMPORANEOUS_RECRUITS}` is used in a variable-substitution context, so you can use variable substitution as well as `$(...)` formula substitution. However, the formulas used in this syntax do not have access to the game state, so I don't believe there's a way to count the number of units satisfying a particular condition (such as having a fire type attack). I think your best bet for something along these lines would be to have a recruit event that stores all matching units to an array (eg `$matching_units`) and then use the length of that array (eg `$matching_units.length`) in your formula.
3. Again, this is a variable substitution context, so you can use variables and substitution formulas. If I recall correctly, WFL has max and min functions, so you should be able to write stuff like `amount="$(max($unit.cost, 60))"` or `amount={ON_DIFFICULTY "max($unit.cost, 60)" "min($unit.cost, 40)" 20}`, or even things like `amount="$(max($unit.cost, {ON_DIFFICULTY 60 40 20}))"`
4. The chance to hit in `{FORCE_CHANCE_TO_HIT}` is also used in a variable substitution context, but I don't think a substitution formula will allow you to access the unit's original chance to hit. As mentioned in the IRC log, it's also a context where you can use WFL _without_ the substitution syntax, which does allow some game state access. However, the only variables available are the units involved with the event (self, other, attacker, defender, and student). There is a `chance_to_hit(unit, loc(x,y))` function in WFL, so you could try using that to get the chance to hit prior to the application of the forced chance… however, since this is in the middle of _altering_ the chance, I can't be certain it'll work.

**Comment by cooljeanius** [comment]
>    1. The recruit amount parameter to `{LIMIT_CONTEMPORANEOUS_RECRUITS}` is used in a variable-substitution context, so you can use variable substitution as well as `$(...)` formula substitution. **However, the formulas used in this syntax do not have access to the game state,** so I don't believe there's a way to count the number of units satisfying a particular condition (such as having a fire type attack). I think your best bet for something along these lines would be to have a recruit event that stores all matching units to an array (eg `$matching_units`) and then use the length of that array (eg `$matching_units.length`) in your formula.
> 

Oh wait... if there's no access to game state there, does this mean that 5f49338e62665d69fd19405199f211c8ff76568d won't actually work like I think it does?

>    2. Again, this is a variable substitution context, so you can use variables and substitution formulas. If I recall correctly, WFL has max and min functions, so you should be able to write stuff like `amount="$(max($unit.cost, 60))"` or `amount={ON_DIFFICULTY "max($unit.cost, 60)" "min($unit.cost, 40)" 20}`, or even things like `amount="$(max($unit.cost, {ON_DIFFICULTY 60 40 20}))"`
> 

Hm, I think I like the middle one there, but... well, so, just to confirm (since I always get my `min` mixed up with my `max`), would using `"max(min($unit.cost, 40), 20)"` for `NORMAL` difficulty work like I'm hoping it does? (i.e. as a constraint to ensure `$unit.cost` is only used if it's between 40 and 20)

>    3. The chance to hit in `{FORCE_CHANCE_TO_HIT}` is also used in a variable substitution context, but I don't think a substitution formula will allow you to access the unit's original chance to hit. As mentioned in the IRC log, it's also a context where you can use WFL _without_ the substitution syntax, which does allow some game state access. However, the only variables available are the units involved with the event (self, other, attacker, defender, and student). There is a `chance_to_hit(unit, loc(x,y))` function in WFL, so you could try using that to get the chance to hit prior to the application of the forced chance… however, since this is in the middle of _altering_ the chance, I can't be certain it'll work.

OK I guess I won't really bother with this one...

**Comment by CelticMinstrel** [comment]
> Oh wait... if there's no access to game state there, does this mean that 5f49338 won't actually work like I think it does?

No, that looks fine. You're not using a substitution formula there; variable substitution works just fine. And if you mean the commented out version, it also looks fine. The formula itself has no access to the game state, but from the formula's perspective, that `$turn_number` has already been substituted (because variable substitution moves from right to left), so it's just a numeric literal.

> Hm, I think I like the middle one there, but... well, so, just to confirm

Your formula for `clamp` looks correct… maybe we should add that as a built-in…

**Comment by cooljeanius** [comment]
ok cool so box 3 is done

**Comment by CelticMinstrel** [comment]
Um… that's not going to work. It's not enclosed in the `$(...)` formula substitution syntax.

**Comment by cooljeanius** [comment]
> Um… that's not going to work. It's not enclosed in the `$(...)` formula substitution syntax.

does that go inside of or outside of the `{ON_DIFFICULTY}` macro?

**Comment by CelticMinstrel** [comment]
It doesn't matter. I'd probably put it inside, just because I think it looks a bit neater that way, but since the hard-mode value is just a number, it'll work equally well outside.

**Comment by cooljeanius** [comment]
OK, so current status of this issue is: box 3 is done, box 1 probably actually isn't necessary, and boxes 2 and 4 are probably too hard for me to bother with. I could theoretically close this now (as partially fixed and partially WONTFIX), but I think I'll leave it open in case I'm feeling inspired later.
